DEV/org.openl.rules.util/test/org/openl/rules/util/StatisticsTest.java

```diff
@@ -86,4 +87,30 @@ public void testAvg() {
         assertEquals(new BigDecimal("5.333333333333333333333333333333333", MathContext.DECIMAL128),
             avg(BigInteger.valueOf(3), BigInteger.valueOf(5), BigInteger.valueOf(8)));
     }
+
+    @Test
+    public void testStandardPopulationDeviation() {
+        assertNull(standardPopulationDeviation(null));
+        assertNull(standardPopulationDeviation(new Double[0]));
+        assertNull(standardPopulationDeviation(new Double[] { null, null }));
+
+        assertEquals(Double.valueOf(4.0276819911981905), standardPopulationDeviation(1, 10, 9));
+        assertEquals(Double.valueOf(0.0), standardPopulationDeviation(9.5));
+        assertEquals(Double.valueOf(3), standardPopulationDeviation(8, null, 2));
+        assertEquals(Double.valueOf(7.788880963698615), standardPopulationDeviation(-10.0, 6.0, 7.0));
+
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation((byte) 3, (byte) 4, (byte) 8));
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation((short) 3, (short) 4, (short) 8));
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation(3, 4, 8));
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation(3L, 4L, 8L));
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation(3f, 4f, 8f));
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation(3d, 4d, 8d));
+        assertEquals(Double.valueOf(2.160246899469287), standardPopulationDeviation((byte) 3, (short) 4, 8));
+        assertEquals(Double.valueOf(2.160246899469287),
+            standardPopulationDeviation(BigInteger.valueOf(3), BigInteger.valueOf(4), BigInteger.valueOf(8)));
+        assertEquals(Double.valueOf(0.5), standardPopulationDeviation(BigDecimal.valueOf(3), BigDecimal.valueOf(4)));
+        assertEquals(Double.valueOf(2.0548046676563256),
+            standardPopulationDeviation(BigInteger.valueOf(3), BigInteger.valueOf(5), BigInteger.valueOf(8)));

```

Comment:
It's believed that using DELTA (or tolerance) is a good practice for floating-point comparisons.

```
private static final double DELTA = 1e-9;
```

The reason why a delta (or tolerance) must be used when comparing floating-point numbers in tests is due to the nature of how these numbers are represented and computed in computers.

Floating-point numbers are represented in computers using a binary format that often can't exactly represent numbers that are simple and exact in decimal form. This is akin to how the fraction `1/3` cannot be precisely represented in decimal format and is approximated to `0.3333`... Similarly, a number like 0.1 cannot be precisely represented in binary floating-point format.

When you perform arithmetic operations with floating-point numbers, these small representation errors can accumulate, leading to results that are slightly off from what you might mathematically expect if the numbers were represented exactly. This phenomenon is known as "floating-point error" or "rounding error."
For example:
```
assertEquals(Double.valueOf(0.3d), Double.valueOf(0.1d + 0.2d));
```

This will often result in False because the actual sum of 0.1 and 0.2 in floating-point arithmetic may be something like 0.30000000000000004 instead of exactly 0.3.

When writing tests, if you compare the expected result with the actual result using strict equality, the test might fail even though the difference is just due to floating-point error and is not significant for most practical purposes.

To account for this, a delta value is used to specify a range within which the actual result can differ from the expected result and still be considered equal. This means that rather than checking for exact equality, you're checking if the actual result is "close enough" to the expected result:

```
assertEquals(Double.valueOf(0.3d), Double.valueOf(0.1d + 0.2d), DELTA);
```

Here, `DELTA` is a small number that defines the acceptable margin of error. If actual is within `±DELTA` of `expected`, the test will pass, allowing for the small inaccuracies inherent in floating-point arithmetic.

Reply:
The implemented functions should return exact values, but not approximated. So using DELTA in assertions is superfluous. Using DELTA will hide issues with rounding and overflowing of numbers.